src/methods.jl

```diff
@@ -10,318 +19,446 @@ derivatives.
+        # `Int128`. At the end we go to `Float64` for fast floating point math, rather than
+        # rational math.
         C = [Rational{Int128}(g^i) for i in 0:(p - 1), g in grid]
         x = zeros(Rational{Int128}, p)
         x[q + 1] = factorial(q)
         return Float64.(C \ x)
     end
 end
 
+# Estimate the bound on the derivative by amplifying the ∞-norm.
+function _make_default_bound_estimator(; condition::Int=DEFAULT_CONDITION)
+    return (f, x) -> condition * maximum(abs, f(x))
+end
 
-# Estimate the bound on the function value and its derivatives at a point
-_estimate_bound(x, cond) = add_tiny(cond * maximum(abs, x))
+function Base.show(io::IO, m::MIME"text/plain", x::FiniteDifferenceMethod)
+    @printf io "FiniteDifferenceMethod:\n"
+    @printf io "  order of method:       %d\n" length(x.grid)
+    @printf io "  order of derivative:   %d\n" x.q
+    @printf io "  grid:                  %s\n" x.grid
+    @printf io "  coefficients:          %s\n" x.coefs
+end
 
 """
-    fdm(m::FiniteDifferenceMethod, f, x[, Val(false)]; kwargs...) -> Real
-    fdm(m::FiniteDifferenceMethod, f, x, Val(true); kwargs...) -> Tuple{FiniteDifferenceMethod, Real}
-
-Compute the derivative of `f` at `x` using the finite differencing method `m`.
-The optional `Val` argument dictates whether the method should be returned alongside the
-derivative value, which can be useful for examining the step size used and other such
-parameters.
-
-The recognized keywords are:
+    function estimate_step(
+        m::FiniteDifferenceMethod,
+        f::Function,
+        x::T;
+        factor::Int=1,
+        max_step=convert(T, 0.1)
+    ) where T<:AbstractFloat
+
+Estimate the step size for a finite difference method `m`. Also estimates the error of the
+estimate of the derivative.
+
+# Arguments
+- `m::FiniteDifferenceMethod`: Finite difference method to estimate the step size for.
+- `f::Function`: Function to evaluate the derivative of.
+- `x::T`: Point to estimate the derivative at.
+
+# Keywords
+- `factor::Int=1`. Factor to amplify the estimated round-off error by. This can be used
+    to force a more conservative step size.
+- `max_step=convert(T, 0.1)`: Maximum step size.
+
+# Returns
+- `Tuple{T, T}`: Estimated step size and an estimate of the error of the finite difference
+    estimate.
+"""
+function estimate_step(
+    m::FiniteDifferenceMethod,
+    f::Function,
+    x::T;
+    factor::Int=1,
+    max_step=convert(T, 0.1)
+) where T<:AbstractFloat
+    p = length(m.coefs)
+    q = m.q
+    f_x = float(f(x))
 
-* `adapt`: The number of adaptive steps to use improve the estimate of `bound`.
-* `bound`: Bound on the value of the function and its derivatives at `x`.
-* `condition`: The condition number. See [`DEFAULT_CONDITION`](@ref).
-* `eps`: The assumed roundoff error. Defaults to `add_tiny(eps())`.
+    # Estimate the bound and round-off error.
+    ε = add_tiny(maximum(eps, f_x)) * factor
```

Comment:
`f_x` must be a scalar so no need for `maximum` here

```suggestion
    ε = add_tiny(eps(f_x)) * factor
```

Reply:
It seems like you didn't resolve this one?

Reply:
Oh i see you other comment sayws we do need this